File: options_trader/risk/trailing_stop.py
from __future__ import annotations

from models.types import ExitSignal, Position


def _clamp(value: float, lower: float, upper: float) -> float:
    return max(lower, min(value, upper))


def update_high_water(position: Position) -> None:
    position.high_water_price = max(position.high_water_price, position.current_price)
    position.high_water_unrealized_pct = max(position.high_water_unrealized_pct, position.unrealized_pct)


def evaluate_trailing_stop(position: Position, risk_config: dict[str, float]) -> ExitSignal:
    update_high_water(position)

    activation = float(risk_config.get("trail_activation_pct", 0.15))
    if position.high_water_unrealized_pct < activation:
        return ExitSignal(False, None, position.unrealized_pct, activation, "Trailing stop not activated.")

    base_trail = float(risk_config.get("base_trail_pct", 0.20))
    dte_scale = (position.dte_remaining / max(position.dte_initial, 1)) if position.dte_initial > 0 else 1.0

    iv = max(position.implied_volatility, 0.01)
    iv_ref = 0.40
    raw_vol_scale = iv / iv_ref
    vol_scale = _clamp(
        raw_vol_scale,
        float(risk_config.get("volatility_scale_min", 0.75)),
        float(risk_config.get("volatility_scale_max", 1.25)),
    )

    trail_pct = max(0.01, base_trail * dte_scale * vol_scale)
    stop_price = position.high_water_price * (1.0 - trail_pct)

    if position.current_price <= stop_price:
        return ExitSignal(
            should_exit=True,
            reason="trailing_stop",
            trigger_value=position.current_price,
            threshold_value=stop_price,
            detail="Trailing stop hit after profit activation.",
        )

    return ExitSignal(
        should_exit=False,
        reason=None,
        trigger_value=position.current_price,
        threshold_value=stop_price,
        detail="Trailing stop active.",
    )

File: options_trader_node/src/risk/trailingStop.js
import { makeExitSignal } from "../models/types.js";

function clamp(value, lower, upper) {
  return Math.max(lower, Math.min(value, upper));
}

export function evaluateTrailingStop(position, riskCfg) {
  position.highWaterPrice = Math.max(position.highWaterPrice ?? position.currentPrice, position.currentPrice);
  position.highWaterUnrealizedPct = Math.max(
    position.highWaterUnrealizedPct ?? position.unrealizedPct,
    position.unrealizedPct,
  );

  const activation = Number(riskCfg.trailActivationPct ?? 0.15);
  if (position.highWaterUnrealizedPct < activation) {
    return makeExitSignal({
      triggerValue: position.unrealizedPct,
      thresholdValue: activation,
      detail: "Trailing stop not activated.",
    });
  }

  const baseTrail = Number(riskCfg.baseTrailPct ?? 0.2);
  const dteScale = position.dteInitial > 0 ? position.dteRemaining / Math.max(position.dteInitial, 1) : 1;
  const volScale = clamp(
    (position.impliedVolatility || 0.4) / 0.4,
    Number(riskCfg.volatilityScaleMin ?? 0.75),
    Number(riskCfg.volatilityScaleMax ?? 1.25),
  );
  const trailPct = Math.max(0.01, baseTrail * dteScale * volScale);
  const stopPrice = position.highWaterPrice * (1 - trailPct);

  if (position.currentPrice <= stopPrice) {
    return makeExitSignal({
      shouldExit: true,
      reason: "trailing_stop",
      triggerValue: position.currentPrice,
      thresholdValue: stopPrice,
      detail: "Trailing stop hit.",
    });
  }

  return makeExitSignal({
    triggerValue: position.currentPrice,
    thresholdValue: stopPrice,
    detail: "Trailing stop active.",
  });
}
